giskard/models/base/model.py

```diff
@@ -593,3 +601,151 @@ def __str__(self) -> str:
+            MetricTool.default_name: MetricTool(model=self, dataset=dataset),
+            SHAPExplanationTool.default_name: SHAPExplanationTool(model=self, dataset=dataset),
+            IssuesScannerTool.default_name: IssuesScannerTool(scan_report=scan_report),
+        }
+
+    @staticmethod
+    def _gather_context(message_list: list) -> str:
+        """Gather context into a single string.
+
+        Given the list of OpenAI's messages, extracts and joins their contents into a single string.
+
+        Parameters
+        ----------
+        message_list : list
+            The list of OpenAI's messages.
+
+        Returns
+        -------
+        str
+            The string with the joined messages' contents.
+        """
+        return "\n".join([json.dumps(asdict(msg)) for msg in message_list])
+
+    def talk(self, question: str, dataset: Dataset, scan_report: ScanReport = None, context: str = ""):
+        """Perform the 'talk' to the model.
+
+        Given `question`, allows to ask the model about prediction result, explanation, model performance, issues, etc.
+
+        Parameters
+        ----------
+        question : str
+            User input query.
+        dataset : Dataset
+            Giskard Dataset to be analysed by the 'talk'.
+        context : str
+            Context of the previous 'talk' results. Necessary to keep context between sequential 'talk' calls.
+        scan_report : ScanReport
+            Giskard Scan Report to be analysed by the 'talk'.
+
+        Returns
+        -------
+        TalkResult
+            The response for the user's prompt.
+        """
+        try:
+            from ...llm.client.copilot import GiskardCopilotClient, ToolChatMessage
+            from ...llm.talk.config import (
+                ERROR_RESPONSE,
+                MODEL_INSTRUCTION,
+                SUMMARY_PROMPT,
+                TALK_CLIENT_CONFIG,
+            )
+            from ..talk_result import TalkResult
+        except ImportError as err:
+            raise GiskardInstallationError(flavor="talk") from err
+
+        client = GiskardCopilotClient()
+
+        available_tools = self._get_available_tools(dataset, scan_report)
+
+        system_prompt = MODEL_INSTRUCTION.format(
+            tools_description="\n".join([tool.description for tool in list(available_tools.values())]),
+            model_name=self.meta.name,
+            model_description=self.meta.description,
+            feature_names=self.meta.feature_names,
+            context=context,
+        )
+
+        messages = [
+            ToolChatMessage(role="system", content=system_prompt),
+            ToolChatMessage(role="user", content=question),
+        ]
+        response = client.complete(
+            messages=messages,
+            tools=[tool.specification for tool in list(available_tools.values())],
+            tool_choice="auto",
+            **TALK_CLIENT_CONFIG,
+        )
+
+        if content := response.content:
```

Comment:
I don't get the use of walrus here

Reply:
yeah good catch, it would've been useful if these where properties with lazy loading of sorts, but no both of these are just simple attributes. I fixed it.